src/wallet/wallet.cpp

```diff
@@ -1156,8 +1156,11 @@ void CWallet::BlockUntilSyncedToCurrentChain() {
     // ::ChainActive().Tip(), otherwise put a callback in the validation interface queue and wait
     // for the queue to drain enough to execute it (indicating we are caught up
     // at least with the time we entered this function).
-    uint256 last_block_hash = WITH_LOCK(cs_wallet, return m_last_block_processed);
-    chain().waitForNotificationsIfTipChanged(last_block_hash);
+    bool tipIsSame = false;
+    do {
+        uint256 last_block_hash = WITH_LOCK(cs_wallet, return m_last_block_processed);
+        tipIsSame = chain().waitForNotificationsIfTipIsNotSame(last_block_hash);
```

Comment:
Shouldn't this whole scope happen under the cs_wallet lock?

Also, shouldn't `m_chain_notifications_handler.reset()` happen inside here under the wallet lock?

Otherwise a new block could come in and put new notifications in the handler while we reset it.

Reply:
> Shouldn't this whole scope happen under the cs_wallet lock?
No, we leave mutex for process thread to finish.
> Also, shouldn't m_chain_notifications_handler.reset() happen inside here under the wallet lock?
No.
> Otherwise a new block could come in and put new notifications in the handler while we reset it.
I'm aware of that, we can have a function that takes a mutex and reset notifications